File: python/decodeforge/q8.py
"""Bit-exact, dependency-free reference semantics for ``DFQ8_B32_V1``.

The module deliberately works with IEEE-754 binary32 bit words.  Keeping the
bits at the API boundary makes signed zeroes and NaN payloads observable and
prevents a host tensor library (or a platform rounding mode) from becoming
part of the format contract.

The reference is intentionally scalar and fairly conservative.  It is an
oracle for quantizers and generated kernels, not a high-throughput inference
implementation.
"""

from __future__ import annotations

import math
import struct
from collections.abc import Sequence
from typing import TypeAlias

FORMAT = "DFQ8_B32_V1"
NUMERIC_MODE = "strict_f32_v1"
BLOCK_SIZE = 32
Q_MIN = -127
Q_MAX = 127
U32_MAX = 0xFFFFFFFF
# This is the largest K for which the predeclared comparator's t*u is < 1.
# Quantization itself accepts every positive-u32 K that can be represented by
# the derived storage checks below; this cap belongs only to comparator gamma.
MAX_COMPARATOR_K = 8_134_399
U = 2.0**-24

SIGN_MASK = 0x80000000
EXP_MASK = 0x7F800000
FRAC_MASK = 0x007FFFFF
POSITIVE_MASK = 0x7FFFFFFF
F32_MAX_BITS = 0x7F7FFFFF
F32_ONE_BITS = 0x3F800000
F32_127_BITS = 0x42FE0000

Bits: TypeAlias = int
BitSequence: TypeAlias = Sequence[int]


class Q8Error(ValueError):
    """A stable semantic rejection with a machine-readable diagnostic."""

    def __init__(
        self,
        code: str,
        summary: str,
        context: dict[str, object] | None = None,
    ) -> None:
        self.code = code
        self.summary = summary
        self.context = dict(context or {})
        self.diagnostic = {
            "schema_version": 1,
            "code": code,
            "severity": "error",
            "component": "quant",
            "summary": summary,
            "context": self.context,
        }
        super().__init__(f"{code}: {summary}")


QuantizationError = Q8Error
ComparatorError = Q8Error
FixtureError = Q8Error


def _error(code: str, summary: str, **context: object) -> Q8Error:
    return Q8Error(code, summary, context)


def _validate_u32(value: object, name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise _error(
            "DFE-QUANT-001", "A dimension or bit word is not an integer.", field=name
        )
    if value < 0 or value > U32_MAX:
        raise _error(
            "DFE-QUANT-001",
            "A dimension or bit word is outside the unsigned binary32 range.",
            field=name,
            value=value,
        )
    return value


def _validate_bits(value: object, name: str = "bits") -> int:
    return _validate_u32(value, name)


def is_finite_f32_bits(bits: int) -> bool:
    """Return whether *bits* encodes a finite binary32 value."""

    bits = _validate_bits(bits)
    return (bits & EXP_MASK) != EXP_MASK


def f32_bits_to_float(bits: int) -> float:
    """Decode one validated binary32 bit word to a Python float."""

    bits = _validate_bits(bits)
    return float(struct.unpack("<f", struct.pack("<I", bits))[0])


def float_to_f32_bits(value: float) -> int:
    """Round a finite Python float to binary32 using IEEE round-to-nearest.

    This helper is for constructing test inputs.  Arithmetic in this module
    uses integer/rational operations instead; in particular division never
    passes through this conversion.
    """

    if not isinstance(value, float):
        raise TypeError("value must be a Python float")
    if not math.isfinite(value):
        raise _error(
            "DFE-QUANT-004", "A source value is not finite.", value=repr(value)
        )
    try:
        return int(struct.unpack("<I", struct.pack("<f", value))[0])
    except (OverflowError, struct.error) as exc:
        raise _error(
            "DFE-QUANT-004",
            "A source value is outside binary32 range.",
            value=repr(value),
        ) from exc


def _components(bits: int) -> tuple[int, int, int, bool]:
    """Return (sign, significand, power-of-two exponent, is_zero)."""

    bits = _validate_bits(bits)
    sign = -1 if bits & SIGN_MASK else 1
    exponent = (bits & EXP_MASK) >> 23
    fraction = bits & FRAC_MASK
    if exponent == 0:
        return sign, fraction, -149, fraction == 0
    if exponent == 0xFF:
        raise ValueError("non-finite binary32 value has no finite components")
    return sign, 0x800000 | fraction, exponent - 150, False


def _round_integer_ratio(numerator: int, denominator: int) -> int:
    """Round a non-negative rational to an integer, ties-to-even."""

    if numerator < 0 or denominator <= 0:
        raise ValueError("invalid non-negative rational")
    quotient, remainder = divmod(numerator, denominator)
    twice = remainder << 1
    if twice > denominator or (twice == denominator and quotient & 1):
        return quotient + 1
    return quotient


def _round_scaled(numerator: int, denominator: int, shift: int) -> int:
    """Return RNE(numerator / denominator * 2**shift)."""

    if shift >= 0:
        numerator <<= shift
    else:
        denominator <<= -shift
    return _round_integer_ratio(numerator, denominator)


def _floor_log2_ratio(numerator: int, denominator: int) -> int:
    """Return floor(log2(numerator / denominator)) for positive integers."""

    if numerator <= 0 or denominator <= 0:
        raise ValueError("log2 ratio requires positive integers")
    exponent = numerator.bit_length() - denominator.bit_length()
    if exponent >= 0:
        if numerator < (denominator << exponent):
            exponent -= 1
    elif (numerator << -exponent) < denominator:
        exponent -= 1
    return exponent


def _round_rational_f32(
    numerator: int,
    denominator: int = 1,
    binary_exponent: int = 0,
    sign: int = 1,
) -> int:
    """Round ``sign * numerator / denominator * 2**binary_exponent`` to f32.

    The implementation handles normal and gradual-underflow results directly
    as integer ratios.  It is therefore suitable for exact midpoint tests and
    does not depend on Python's binary64 division.
    """

    if numerator < 0 or denominator <= 0 or sign not in {-1, 1}:
        raise ValueError("invalid rational f32 arguments")
    sign_bit = SIGN_MASK if sign < 0 else 0
    if numerator == 0:
        return sign_bit

    exponent = _floor_log2_ratio(numerator, denominator) + binary_exponent
    if exponent < -126:
        # A subnormal unit is 2**-149.  Rounding can carry exactly to the
        # smallest normal, which is represented by exponent field one.
        subnormal = _round_scaled(numerator, denominator, binary_exponent + 149)
        if subnormal == 0:
            return sign_bit
        if subnormal >= 1 << 23:
            return sign_bit | (1 << 23)
        return sign_bit | subnormal

    # A normal has 24 significant bits, with the leading bit implicit.
    significand = _round_scaled(
        numerator,
        denominator,
        binary_exponent - exponent + 23,
    )
    if significand >= 1 << 24:
        significand >>= 1
        exponent += 1
    if exponent > 127:
        return sign_bit | EXP_MASK
    if exponent < -126:  # defensive: the subnormal branch should handle this
        subnormal = _round_scaled(numerator, denominator, binary_exponent + 149)
        if subnormal == 0:
            return sign_bit
        if subnormal >= 1 << 23:
            return sign_bit | (1 << 23)
        return sign_bit | subnormal
    return sign_bit | ((exponent + 127) << 23) | (significand - (1 << 23))


def _finite_components(bits: int) -> tuple[int, int, int, bool]:
    if not is_finite_f32_bits(bits):
        raise ValueError("binary32 operand is not finite")
    return _components(bits)


def f32_mul_bits(left: int, right: int) -> int:
    """Multiply two finite binary32 values and round once to binary32."""

    sign_left, num_left, power_left, zero_left = _finite_components(left)
    sign_right, num_right, power_right, zero_right = _finite_components(right)
    sign = sign_left * sign_right
    if zero_left or zero_right:
        return SIGN_MASK if sign < 0 else 0
    return _round_rational_f32(
        num_left * num_right,
        binary_exponent=power_left + power_right,
        sign=sign,
    )


def f32_add_bits(left: int, right: int) -> int:
    """Add two finite binary32 values and round once to binary32."""

    sign_left, num_left, power_left, zero_left = _finite_components(left)
    sign_right, num_right, power_right, zero_right = _finite_components(right)
    if zero_left and zero_right:
        # IEEE round-to-nearest chooses +0 for opposite-signed zeroes and for
        # an exact cancellation.  Preserve -0 only when both operands are -0.
        if sign_left < 0 and sign_right < 0:
            return SIGN_MASK
        return 0
    if zero_left:
        return right
    if zero_right:
        return left

    common_power = min(power_left, power_right)
    signed_left = sign_left * (num_left << (power_left - common_power))
    signed_right = sign_right * (num_right << (power_right - common_power))
    total = signed_left + signed_right
    if total == 0:
        return 0
    return _round_rational_f32(
        abs(total),
        binary_exponent=common_power,
        sign=1 if total > 0 else -1,
    )


def f32_div_bits(numerator_bits: int, denominator_bits: int) -> int:
    """Divide finite binary32 values with exact rational RNE semantics."""

    sign_numerator, num, power_num, zero_num = _finite_components(numerator_bits)
    sign_denominator, den, power_denominator, zero_denominator = _finite_components(
        denominator_bits
    )
    if zero_denominator:
        raise ZeroDivisionError("binary32 division by zero")
    sign = sign_numerator * sign_denominator
    if zero_num:
        return SIGN_MASK if sign < 0 else 0
    return _round_rational_f32(
        num,
        den,
        power_num - power_denominator,
        sign,
    )


def f32_from_int(value: int) -> int:
    """Return the exactly rounded binary32 representation of an integer."""

    if isinstance(value, bool) or not isinstance(value, int):
        raise TypeError("value must be an integer")
    if value == 0:
        return 0
    return _round_rational_f32(abs(value), sign=1 if value > 0 else -1)

File: compiler/decodeforge-core/src/q8.rs
//! Exact binary32 primitives for `DFQ8_B32_V1`.
//!
//! The public API intentionally deals in binary32 bit words.  This keeps
//! signed zeroes and NaN payloads observable and makes the Rust oracle
//! independent of a tensor library or a host language's floating point
//! conversions.  This layer supplies host-independent integer/rational
//! arithmetic; Q8 storage and evaluation are introduced separately.

use num_bigint::{BigInt, BigUint, Sign};
use num_traits::{One, ToPrimitive, Zero};
use serde::Serialize;
use serde_json::Value;
use std::collections::BTreeMap;
use std::fmt;

/// The frozen Q8 storage format.
pub const FORMAT: &str = "DFQ8_B32_V1";
/// The frozen scalar arithmetic mode.
pub const NUMERIC_MODE: &str = "strict_f32_v1";
/// Number of lanes in one physical Q8 block.
pub const BLOCK_SIZE: usize = 32;
/// Smallest representable signed Q8 value (negative 128 is forbidden).
pub const Q_MIN: i16 = -127;
/// Largest representable signed Q8 value.
pub const Q_MAX: i16 = 127;

const SIGN_MASK: u32 = 0x8000_0000;
const EXP_MASK: u32 = 0x7f80_0000;
const FRAC_MASK: u32 = 0x007f_ffff;
const POSITIVE_MASK: u32 = 0x7fff_ffff;

/// A stable semantic rejection with a machine-readable diagnostic code.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Q8Error {
    /// Stable catalog code (for example `DFE-QUANT-003`).
    pub code: &'static str,
    /// Frozen human-readable summary.
    pub summary: &'static str,
    /// Small, deterministic context map useful to callers and diagnostics.
    pub context: BTreeMap<String, Value>,
}
impl Q8Error {
    fn new(code: &'static str, summary: &'static str) -> Self {
        Self {
            code,
            summary,
            context: BTreeMap::new(),
        }
    }

    /// Render the closed diagnostic object as canonical JSON.
    pub fn diagnostic_json(&self) -> String {
        #[derive(Serialize)]
        struct Diagnostic<'a> {
            schema_version: u32,
            code: &'a str,
            severity: &'static str,
            component: &'static str,
            summary: &'a str,
            context: &'a BTreeMap<String, Value>,
        }
        serde_json::to_string(&Diagnostic {
            schema_version: 1,
            code: self.code,
            severity: "error",
            component: component_for_code(self.code),
            summary: self.summary,
            context: &self.context,
        })
        .expect("diagnostic values serialize")
    }
}

impl fmt::Display for Q8Error {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}: {}", self.code, self.summary)
    }
}

impl std::error::Error for Q8Error {}

fn quant_error(code: &'static str, summary: &'static str) -> Q8Error {
    Q8Error::new(code, summary)
}

fn component_for_code(code: &str) -> &'static str {
    if code.starts_with("DFE-SCHEMA-") {
        "schema"
    } else {
        "quant"
    }
}

/// Whether a bit word encodes a finite binary32 number.
pub const fn is_finite_f32_bits(bits: u32) -> bool {
    (bits & EXP_MASK) != EXP_MASK
}

/// Decode a validated binary32 word to the host's f32 value.
pub fn f32_bits_to_float(bits: u32) -> f32 {
    f32::from_bits(bits)
}

fn round_integer_ratio(numerator: &BigUint, denominator: &BigUint) -> BigUint {
    debug_assert!(!denominator.is_zero());
    let quotient = numerator / denominator;
    let remainder = numerator % denominator;
    let twice = &remainder << 1usize;
    if twice > *denominator || (twice == *denominator && quotient.bit(0)) {
        quotient + BigUint::one()
    } else {
        quotient
    }
}

fn round_scaled(numerator: &BigUint, denominator: &BigUint, shift: i32) -> BigUint {
    let mut numerator = numerator.clone();
    let mut denominator = denominator.clone();
    if shift >= 0 {
        numerator <<= usize::try_from(shift).expect("non-negative f32 shift fits usize");
    } else {
        denominator <<= usize::try_from(-shift).expect("negative f32 shift fits usize");
    }
    round_integer_ratio(&numerator, &denominator)
}

fn floor_log2_ratio(numerator: &BigUint, denominator: &BigUint) -> i32 {
    debug_assert!(!numerator.is_zero() && !denominator.is_zero());
    let mut exponent = i32::try_from(numerator.bits()).expect("binary32 ratio fits i32")
        - i32::try_from(denominator.bits()).expect("binary32 ratio fits i32");
    if exponent >= 0 {
        let shifted = denominator << usize::try_from(exponent).expect("f32 exponent fits usize");
        if numerator < &shifted {
            exponent -= 1;
        }
    } else {
        let shifted = numerator << usize::try_from(-exponent).expect("f32 exponent fits usize");
        if shifted < *denominator {
            exponent -= 1;
        }
    }
    exponent
}

/// Round a signed rational binary value to one IEEE-754 binary32 word.
///
/// All intermediate arithmetic here is integer arithmetic.  This mirrors the
/// Python reference's rational implementation and therefore does not depend
/// on the host FPU, floating-point environment, contraction, or extended
/// precision.  The normal and gradual-underflow paths include the carry into
/// the next exponent and overflow to infinity.
fn round_rational_f32(
    numerator: BigUint,
    denominator: BigUint,
    binary_exponent: i32,
    sign: i8,
) -> u32 {
    debug_assert!(sign == -1 || sign == 1);
    debug_assert!(!denominator.is_zero());
    let sign_bit = if sign < 0 { SIGN_MASK } else { 0 };
    if numerator.is_zero() {
        return sign_bit;
    }

    let mut exponent = floor_log2_ratio(&numerator, &denominator) + binary_exponent;
    if exponent < -126 {
        let subnormal = round_scaled(&numerator, &denominator, binary_exponent + 149);
        if subnormal.is_zero() {
            return sign_bit;
        }
        let normal_threshold = BigUint::one() << 23usize;
        if subnormal >= normal_threshold {
            return sign_bit | (1 << 23);
        }
        return sign_bit | subnormal.to_u32().expect("binary32 subnormal fits u32");
    }

    let mut significand = round_scaled(&numerator, &denominator, binary_exponent - exponent + 23);
    let significand_threshold = BigUint::one() << 24usize;
    if significand >= significand_threshold {
        significand >>= 1usize;
        exponent += 1;
    }
    if exponent > 127 {
        return sign_bit | EXP_MASK;
    }
    if exponent < -126 {
        let subnormal = round_scaled(&numerator, &denominator, binary_exponent + 149);
        if subnormal.is_zero() {
            return sign_bit;
        }
        let normal_threshold = BigUint::one() << 23usize;
        if subnormal >= normal_threshold {
            return sign_bit | (1 << 23);
        }
        return sign_bit | subnormal.to_u32().expect("binary32 subnormal fits u32");
    }
    let hidden_bit = BigUint::one() << 23usize;
    let fraction = (significand - hidden_bit)
        .to_u32()
        .expect("binary32 significand fits u32");
    sign_bit | (((exponent + 127) as u32) << 23) | fraction
}

fn components(bits: u32) -> (i8, BigUint, i32, bool) {
    let sign = if bits & SIGN_MASK == 0 { 1 } else { -1 };
    let exponent_field = (bits & EXP_MASK) >> 23;
    let fraction = bits & FRAC_MASK;
    if exponent_field == 0 {
        return (sign, BigUint::from(fraction), -149, fraction == 0);
    }
    debug_assert_ne!(exponent_field, 0xff);
    (
        sign,
        BigUint::from((1u32 << 23) | fraction),
        exponent_field as i32 - 150,
        false,
    )
}

fn finite_components(bits: u32) -> (i8, BigUint, i32, bool) {
    debug_assert!(is_finite_f32_bits(bits));
    components(bits)
}

/// Multiply two finite binary32 values, rounding once to binary32.
///
/// The operation is implemented with integer/rational arithmetic, including
/// signed zero and gradual underflow, so no native floating-point operation
/// can influence oracle bits.
pub fn f32_mul_bits(left: u32, right: u32) -> Result<u32, Q8Error> {
    if !is_finite_f32_bits(left) || !is_finite_f32_bits(right) {
        return Err(quant_error(
            "DFE-QUANT-010",
            "A binary32 arithmetic operand is not finite.",
        ));
    }
    let (sign_left, numerator_left, power_left, zero_left) = finite_components(left);
    let (sign_right, numerator_right, power_right, zero_right) = finite_components(right);
    let sign = sign_left * sign_right;
    if zero_left || zero_right {
        return Ok(if sign < 0 { SIGN_MASK } else { 0 });
    }
    Ok(round_rational_f32(
        numerator_left * numerator_right,
        BigUint::one(),
        power_left + power_right,
        sign,
    ))
}

/// Add two finite binary32 values, rounding once to binary32.
pub fn f32_add_bits(left: u32, right: u32) -> Result<u32, Q8Error> {
    if !is_finite_f32_bits(left) || !is_finite_f32_bits(right) {
        return Err(quant_error(
            "DFE-QUANT-010",
            "A binary32 arithmetic operand is not finite.",
        ));
    }
    let (sign_left, numerator_left, power_left, zero_left) = finite_components(left);
    let (sign_right, numerator_right, power_right, zero_right) = finite_components(right);
    if zero_left && zero_right {
        return Ok(if sign_left < 0 && sign_right < 0 {
            SIGN_MASK
        } else {
            0
        });
    }
    if zero_left {
        return Ok(right);
    }
    if zero_right {
        return Ok(left);
    }

    let common_power = power_left.min(power_right);
    let mut signed_left = BigInt::from_biguint(
        Sign::Plus,
        numerator_left
            << usize::try_from(power_left - common_power).expect("f32 exponent fits usize"),
    );
    let mut signed_right = BigInt::from_biguint(
        Sign::Plus,
        numerator_right
            << usize::try_from(power_right - common_power).expect("f32 exponent fits usize"),
    );
    if sign_left < 0 {
        signed_left = -signed_left;
    }
    if sign_right < 0 {
        signed_right = -signed_right;
    }
    let total = signed_left + signed_right;
    if total.is_zero() {
        return Ok(0);
    }
    Ok(round_rational_f32(
        total.magnitude().clone(),
        BigUint::one(),
        common_power,
        if total.sign() == Sign::Minus { -1 } else { 1 },
    ))
}

/// Divide two finite binary32 values, rounding once to binary32.
pub fn f32_div_bits(numerator: u32, denominator: u32) -> Result<u32, Q8Error> {
    if !is_finite_f32_bits(numerator) || !is_finite_f32_bits(denominator) {
        return Err(quant_error(
            "DFE-QUANT-006",
            "A scale or division operand is not finite.",
        ));
    }
    if denominator & POSITIVE_MASK == 0 {
        return Err(quant_error(
            "DFE-QUANT-006",
            "A scale or division operand is zero.",
        ));
    }
    let (sign_numerator, numerator, power_numerator, zero_numerator) = finite_components(numerator);
    let (sign_denominator, denominator, power_denominator, zero_denominator) =
        finite_components(denominator);
    let sign = sign_numerator * sign_denominator;
    if zero_numerator {
        return Ok(if sign < 0 { SIGN_MASK } else { 0 });
    }
    debug_assert!(!zero_denominator);
    Ok(round_rational_f32(
        numerator,
        denominator,
        power_numerator - power_denominator,
        sign,
    ))
}

/// Convert a small integer exactly to binary32.
pub fn f32_from_int(value: i32) -> u32 {
    if value == 0 {
        return 0;
    }
    round_rational_f32(
        BigUint::from(value.unsigned_abs()),
        BigUint::one(),
        0,
        if value < 0 { -1 } else { 1 },
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn arithmetic_and_subnormal_boundaries_match_contract() {
        assert_eq!(f32_div_bits(0x0000_003f, 0x42fe_0000).unwrap(), 0);
        assert_eq!(f32_div_bits(0x0000_0040, 0x42fe_0000).unwrap(), 1);
        assert_eq!(f32_mul_bits(0x3f80_0000, 0xbf00_0000).unwrap(), 0xbf00_0000);
        assert_eq!(f32_add_bits(0x3f80_0000, 0x3400_0000).unwrap(), 0x3f80_0001);
    }

    #[test]
    fn arithmetic_matches_python_golden_vectors() {
        let additions = [
            (0x0000_0000, 0x8000_0000, 0x0000_0000),
            (0x8000_0000, 0x8000_0000, 0x8000_0000),
            (0x3f80_0000, 0xbf80_0000, 0x0000_0000),
            (0x007f_ffff, 0x0000_0001, 0x0080_0000),
            (0x0080_0000, 0x8080_0000, 0x0000_0000),
            (0x3f80_0000, 0x3400_0000, 0x3f80_0001),
            (0x7f7f_ffff, 0x7f7f_ffff, 0x7f80_0000),
            (0xff7f_ffff, 0xff7f_ffff, 0xff80_0000),
        ];
        for (left, right, expected) in additions {
            assert_eq!(f32_add_bits(left, right).unwrap(), expected);
        }

        let multiplications = [
            (0x8000_0000, 0x3f80_0000, 0x8000_0000),
            (0x0000_0001, 0x3f80_0000, 0x0000_0001),
            (0x0000_0001, 0x3f00_0000, 0x0000_0000),
            (0x3fc0_0000, 0x4020_0000, 0x4070_0000),
            (0x7f7f_ffff, 0x7f7f_ffff, 0x7f80_0000),
            (0xbf80_0000, 0x4040_0000, 0xc040_0000),
        ];
        for (left, right, expected) in multiplications {
            assert_eq!(f32_mul_bits(left, right).unwrap(), expected);
        }

        let divisions = [
            (0x8000_0000, 0x4040_0000, 0x8000_0000),
            (0x0000_0001, 0x3f00_0000, 0x0000_0002),
            (0x0080_0000, 0x4000_0000, 0x0040_0000),
            (0x3f80_0000, 0x4040_0000, 0x3eaa_aaab),
            (0x7f7f_ffff, 0x0080_0000, 0x7f80_0000),
            (0xbf80_0000, 0x3f00_0000, 0xc000_0000),
        ];
        for (numerator, denominator, expected) in divisions {
            assert_eq!(f32_div_bits(numerator, denominator).unwrap(), expected);
        }
    }
}
